packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts

```diff
@@ -0,0 +1,221 @@
+import type { TSESLint, TSESTree } from '@typescript-eslint/utils';
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+
+import * as util from '../util';
+
+const nodeTypesAreEqual = <T extends TSESTree.Node>(
+  actualTypeNode: TSESTree.Node,
+  expectedTypeNode: T,
+): actualTypeNode is T => {
+  return actualTypeNode.type === expectedTypeNode.type;
+};
+
+const constituentsAreEqual = (
+  actualTypeNode: TSESTree.TypeNode,
+  expectedTypeNode: TSESTree.TypeNode,
+): boolean => {
+  switch (expectedTypeNode.type) {
+    // These types should never occur as constituents of a union/intersection
+    case AST_NODE_TYPES.TSAbstractKeyword:
+    case AST_NODE_TYPES.TSAsyncKeyword:
+    case AST_NODE_TYPES.TSDeclareKeyword:
+    case AST_NODE_TYPES.TSExportKeyword:
+    case AST_NODE_TYPES.TSNamedTupleMember:
+    case AST_NODE_TYPES.TSOptionalType:
+    case AST_NODE_TYPES.TSPrivateKeyword:
+    case AST_NODE_TYPES.TSProtectedKeyword:
+    case AST_NODE_TYPES.TSPublicKeyword:
+    case AST_NODE_TYPES.TSReadonlyKeyword:
+    case AST_NODE_TYPES.TSRestType:
+    case AST_NODE_TYPES.TSStaticKeyword:
+    case AST_NODE_TYPES.TSTypePredicate:
+      /* istanbul ignore next */
+      throw new Error(`Unexpected Type ${expectedTypeNode.type}`);
+
+    default:
+      if (nodeTypesAreEqual(actualTypeNode, expectedTypeNode)) {
+        return astNodesAreEquals(actualTypeNode, expectedTypeNode);
+      }
+      return false;
+  }
+};
+
+const astIgnoreKeys = ['range', 'loc', 'parent'];
+
+type NodeElementsType =
+  | number
+  | string
+  | bigint
+  | boolean
+  | null
+  | undefined
+  | RegExp
+  | TSESTree.BaseNode
+  | TSESTree.BaseNode[]
+  | [number, number];
+
+const astNodesAreEquals = <T extends NodeElementsType>(
+  actualNode: T,
+  expectedNode: T,
+): boolean => {
+  if (actualNode === expectedNode) {
+    return true;
+  }
+  if (
+    actualNode &&
+    expectedNode &&
+    typeof actualNode == 'object' &&
+    typeof expectedNode == 'object'
+  ) {
+    if (Array.isArray(actualNode) && Array.isArray(expectedNode)) {
+      if (actualNode.length != expectedNode.length) {
+        return false;
+      }
+      return !actualNode.some(
+        (nodeEle, index) => !astNodesAreEquals(nodeEle, expectedNode[index]),
+      );
+    }
+    if (isAstNodeType(actualNode) && isAstNodeType(expectedNode)) {
+      const actualNodeKeys = Object.keys(actualNode).filter(
+        key => !astIgnoreKeys.includes(key),
+      );
+      const expectedNodeKeys = Object.keys(expectedNode).filter(
+        key => !astIgnoreKeys.includes(key),
+      );
+      if (actualNodeKeys.length === expectedNodeKeys.length) {
+        return !actualNodeKeys.some(
+          actualNodeKey =>
+            !astNodesAreEquals(
+              actualNode[actualNodeKey],
+              expectedNode[actualNodeKey],
+            ),
+        );
+      }
+    }
+  }
+  return false;
+};
```

Comment:
> ... changed to compare ASTNode recursively. Specifically, this rule is looking for an exact match for each ASTNode excluding `range`, `loc`, and `parent` properties ...

This won't work in all cases 😞. Types might be effectively the same but have slightly different AST representations. One of the test cases as-is demonstrates this:

```ts
type T = "1" | '1';
```

Another example is union types with different orders:

```ts
type T = (number | string)[] | (string | number)[];
```

I think this rule is going to have to use the type checker. Users are going to be rather disappointed that this (simplified from your example) doesn't give an error:

```ts
type A = string;
type B = string;
type C = A | B;
```